src/base_alloc/base_alloc.c

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+*/
+
+#include <assert.h>
+
+#include <base_alloc.h>
+#include <base_alloc_linear.h>
+#include <base_alloc_os_mem.h>
+
+// pool size of the base allocator,
+// it will be reallocated if we run out of memory.
+#define BA_POOL_SIZE (ba_os_get_page_size())
+
+// alignment of the base allocator
+#define MEMORY_ALIGNMENT (8)
+
+typedef struct ba_list_item_t ba_list_item_t;
+
+struct ba_list_item_t {
+    ba_list_item_t *next;
+    void *addr;
+};
+
+struct umf_ba_pool_t {
+    size_t aligned_size;
+    ba_list_item_t *free_list;
+    ba_list_item_t *alloc_list;
+#ifdef DEBUG
+    long tid;
+#endif /* DEBUG */
+};
+
+static int ba_fill_pool(umf_ba_pool_t *pool) {
+    assert(pool->free_list == NULL);
+
+    char *pool_memory = ba_os_alloc(BA_POOL_SIZE);
+    if (!pool_memory) {
+        return -1;
+    }
+
+    size_t pool_size_left = BA_POOL_SIZE;
+    ba_list_item_t *last_list_item = NULL;
+
+    while (pool_size_left >= pool->aligned_size) {
+        ba_list_item_t *new_list_item =
+            (ba_list_item_t *)linear_alloc(sizeof(ba_list_item_t));
+        if (!new_list_item) {
+            return -1;
+        }
+
+        new_list_item->next = last_list_item;
+        new_list_item->addr = pool_memory;
+
+        last_list_item = new_list_item;
+        pool_size_left -= pool->aligned_size;
+        pool_memory += pool->aligned_size;
+    }
+
+    pool->free_list = last_list_item;
+
+    return 0;
+}
+
+umf_ba_pool_t *umf_ba_create(size_t size) {
+    size_t aligned_size;
+    // align size to MEMORY_ALIGNMENT bytes
+    size_t rest = size & (MEMORY_ALIGNMENT - 1);
```

Comment:
do we have a helper function for calculating aligned address?

Reply:
Added.